### ISSUE TITLE
Can't bind controls to Caps Lock

### ISSUE THREAD

**Issue by Jayman2000** [dxx-rebirth/dxx-rebirth#409]
### Environment

Both Ubuntu 18.04 packages (d1x-rebirth version 0.58.1-1build1, and d2x-rebirth version 0.58.1-1.1), and versions compiled from commit fe77949a824e23e7237fd5994a06f6ab072ee271
#### Operating System Environment

* Ubuntu
  * [x] Ubuntu 18.04 LTS (Bionic Beaver)

#### CPU environment

* [x] x86\_64 (64-bit Intel/AMD)

### Description

I can't bind any action to Caps Lock.
#### Regression status

Bug is present in version 0.58.1, but not in the original versions of Descent 1/2 running in DOSBox.
### Steps to Reproduce

1. Open the "Customize Keyboard" menu.
2. Select any of the purple rectangles associated with an action.
3. Press Enter.
4. Press Caps Lock.
#### Expected behavior
The purple square stops strobing, and the '?' inside of it is replaced with the text 'CPSLK'. Starting a game and pressing the Caps Lock key results in the action being performed.
#### Observed behavior
The purple square keeps strobing, and the '?' remains. Starting a game and pressing the Caps Lock key causes nothing to happen.
* Engines affected:
  * [x] D1X-Rebirth
  * [x] D2X-Rebirth

**Comment by vLKp** [comment]
I can reproduce this, but only if I do *not* set __-nostickykeys__.  When __-nostickykeys__ is set, I get the behavior you describe as "Expected."  Please confirm that passing this option produces the desired result for you as well.

**Comment by Jayman2000** [comment]
If i pass the **-nostickykeys** flag (in either same Ubuntu package versions, or the same git builds I did), then I get the expected behavior. Once an action is bound to Caps Lock, I can trigger it weather or not I pass the **-nostickykeys** flag.

**Comment by vLKp** [comment]
While it may be surprising to the user that previously configured keys work even without __-nostickykeys__, I think that is not a problem.  The user did configure the key that way, and can remove or avoid that key if they no longer want it.  Since __-nostickykeys__ does enable binding capslock as an action key, I think there is no bug here.  Do you agree?  If not, what do you think should be changed?

**Comment by Jayman2000** [comment]
I'm not sure if I agree or not because I'm not sure what **-nostickykeys** is supposed to do. The only documentation I could find for **-nostickykeys** is that it "[m]ake[s] CapsLock and NumLock non-sticky". This description was too vague to clear up my confusion. What does 'non-sticky' mean in this context.

According to  [Wikipedia](https://en.wikipedia.org/w/index.php?title=Sticky_keys&oldid=862167934), "[s]ticky keys is an accessibility feature [... that allows] the user to press and release a modifier key, such as Shift, Ctrl, Alt, or the Windows key, and have it remain active until any other key is pressed." What DXX-Rebirth is calling sticky keys does not line up with that definition. For one thing, sticky keys is supposed to affect modifier keys, but in DXX-Rebirth it affects Caps Lock and Num Lock. There are no key combos (that I'm aware of) that involve Caps Lock or Num Lock in DXX-Rebirth. Secondly, Caps Lock and Num Lock do not behave the way they Wikipedia says they would when **-nostickykeys** is _not_ passed. Let's say that I bound slide down to Caps Lock. According to Wikipedia, when I press Caps Lock I should continue to slide downwards until I hit another key. In DXX-Rebirth, I continue sliding downwards until I hit Caps Lock again, even if I hit, say, the A key. This behavior might be desirable, but I don't think that "sticky keys" is a very apt way of describing it.

I also do not understand why passing **-nostickykeys** would be required in order to bind something to Caps Lock. Other pieces of software that implement sticky keys do not have this restriction. For example, in Windows 10 there is a setting that causes a bubble to appear around the cursor when the Ctrl key is released (effectively binding an action to Ctrl). Weather or not sticky keys is enabled does not affect the user's ability to enable the cursor location feature, despite the fact that enabling both features produces an odd and potentially undesirable result (the user must press and release Ctrl three times before the bubble appears around their mouse).

In summary,
1. What is DXX-Rebirth's intended definition of sticky keys vs. no sticky keys?
2. What is the rationale behind restricting a user from binding an action to Caps Lock/Num Lock when sticky keys is on?

**Comment by vLKp** [comment]
@zicodxx : could you address these questions?  The relevant code predates my changes.

@Jayman2000 : I must defer to @zicodxx to address your questions.  I have maintained the sticky key semantics as they were when I began working on the code.  I do not have definitive answers to your questions.

**Comment by zicodxx** [comment]
@Jayman2000 @vLKp 
Sorry for the late reply. So in hindsight, yes "nostickykeys" was a bad choice of words here (not a native speaker). And it seems from your description that things may have changed from how they worked before. 
-nostickykeys sets SDL flag SDL_DISABLE_LOCK_KEYS. Without that, Pressing and releasing Capslock once would not trigger key up event, meaning that if you mapped laser firing to it, your lasers would continue firing automatically until you press the key again, removing Caps locking. With the argument (i.e. SDL_DISABLE_LOCK_KEYS enabled) Capslock and Numlock behave like normal keys, triggering down event once pressed and up event once released - independent from the lock status of the key. 
Considering the confusion in this thread, it seems this behaviour has changed over time, so I will need to check whether we still need this option at all.

**Comment by zicodxx** [comment]
I checked this further and the argument (i.e. SDL_DISABLE_LOCK_KEYS) is still necessary to make those keys to behave normally (i.e. releasing the key will actually send KEYUP).

I do think it might be good to introduce the following changes:
- Renaming the argument to -nolockkeys
- Allowing to map those keys, no matter if the argument was given or not (in case someone wants to use capslock for auto fire for example)

Should only be a really small change. LEt me know if you're okay with that @vLKp . :)

**Comment by vLKp** [comment]
I don't use lock keys for anything, so I have no objection.  Just beware that renaming command line options can surprise users who expect the old option to have meaning.
